overlord/snapstate/handlers.go

```diff
@@ -1150,59 +1150,178 @@ func (m *SnapManager) doCopySnapData(t *state.Task, _ *tomb.Tomb) error {
 		return copyDataErr
 	}
 
-	// the migration hasn't been done - do it now
-	if opts.UseHidden && !opts.MigratedToHidden {
-		if err = m.backend.HideSnapData(snapsup.InstanceName()); err != nil {
+	action := triggeredMigration(newInfo.Base, opts)
+	var undo func()
 
-			// undo the migration. In contrast to copy data for which the new revision
-			// directory can just be discarded, the snap data migration introduces
-			// a change that affects all revisions of the snap and thus needs to be reverted
-			if err := m.backend.UndoHideSnapData(snapsup.InstanceName()); err != nil {
-				src := filepath.Join("~", dirs.HiddenSnapDataHomeDir, snapsup.InstanceName())
-				dst := filepath.Join("~", dirs.UserHomeSnapDir, snapsup.InstanceName())
-
-				st.Lock()
-				st.Warnf("cannot undo snap dir hiding (move all user's %s to %s): %v",
-					src, dst, err)
-				st.Unlock()
-			}
+	switch action {
+	case hidden:
+		undo, err = m.doHiddenMigration(t, st, snapsup, newInfo)
+	case revertHidden:
+		undo, err = m.revertHiddenMigration(t, st, snapsup)
+	case home:
+		undo, err = m.doHomeMigration(t, st, snapsup, newInfo.Revision)
+	case full:
+		undo, err = m.doFullMigration(t, st, snapsup, newInfo)
+	default:
+		// for now, don't support reverting the ~/Snap migration in a refresh
+		err = nil
+	}
 
-			return err
+	// if task fails, undo any migration actions taken
+	defer func() {
+		if err != nil && undo != nil {
+			undo()
 		}
+	}()
 
-		snapsup.MigratedHidden = true
-		st.Lock()
-		err := SetTaskSnapSetup(t, snapsup)
-		st.Unlock()
-		if err != nil {
-			return fmt.Errorf("cannot set migration status to done: %w", err)
+	return err
+}
+
+type migration string
+
+const (
+	// none states that no action should be taken
+	none migration = "none"
+	// hidden migrates ~/snap to ~/.snap
+	hidden migration = "hidden"
+	// revertHidden undoes the hidden migration (i.e., moves ~/.snap to ~/snap)
+	revertHidden migration = "revertHidden"
+	// home migrates the new home to ~/Snap
+	home migration = "home"
+	// full migrates ~/snap to ~/.snap and the new home to ~/Snap
+	full migration = "full"
+)
+
+func triggeredMigration(newBase string, opts *dirMigrationOptions) migration {
+	// we're refreshing to a core22 revision
+	if atLeastCore22(newBase) {
+		if opts.MigratedToHidden && !opts.MigratedToExposedHome {
+			// ~/.snap migration already happened so initialize ~/Snap only
+			return home
+		} else if !opts.MigratedToHidden {
```

Comment:
Can this just be `if !opts.MigratedToHidden` (dropping `else`)?

Reply:
Sure. Dropped (and the other as well)